libyul/optimiser/RedundantStoreBase.h

```diff
@@ -0,0 +1,116 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Base class for both RedundantAsssignEliminator and RedundantStoreEliminator.
+ */
+
+#pragma once
+
+#include <libyul/optimiser/ASTWalker.h>
+#include <libyul/AST.h>
+
+#include <range/v3/action/remove_if.hpp>
+
+#include <variant>
+
+
+namespace solidity::yul
+{
+struct Dialect;
+
+/**
+ * Base class for both RedundantAsssignEliminator and RedundantStoreEliminator.
+ */
+class RedundantStoreBase: public ASTWalker
+{
+public:
+	explicit RedundantStoreBase(Dialect const& _dialect): m_dialect(_dialect) {}
+
+	using ASTWalker::operator();
+	void operator()(If const& _if) override;
+	void operator()(Switch const& _switch) override;
+	void operator()(FunctionDefinition const&) override;
+	void operator()(ForLoop const&) override;
+	void operator()(Break const&) override;
+	void operator()(Continue const&) override;
+
+protected:
+	class State
+	{
+	public:
+		enum Value { Unused, Undecided, Used };
+		State(Value _value = Undecided): m_value(_value) {}
+		inline bool operator==(State _other) const { return m_value == _other.m_value; }
+		inline bool operator!=(State _other) const { return !operator==(_other); }
+		static inline void join(State& _a, State const& _b)
+		{
+			// Using "max" works here because of the order of the values in the enum.
+			_a.m_value =  Value(std::max(int(_a.m_value), int(_b.m_value)));
+		}
+	private:
+		Value m_value = Undecided;
+	};
+
+	using TrackedStores = std::map<YulString, std::map<Statement const*, State>>;
```

Comment:
Might make sense to use a compare function on the inner map based on Statement id or something similar

Reply:
You mean to ensure determinism? I don't think this is needed, to be honest. Furthermore, there are no IDs in yul.

Reply:
`YulString` comparison is "weird", but deterministic.